{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Utils_16S_Analysis import *\n",
    "\n",
    "rcParams = {'font.size': 12, 'font.weight': 'normal', 'font.family': 'sans-serif',\n",
    "            'axes.unicode_minus':False, 'axes.labelweight':'normal'}\n",
    "plt.rcParams.update(rcParams)\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "warnings.simplefilter('ignore')\n",
    "\n",
    "from copy import deepcopy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filedir = '/Users/harihara/Mount-2/hotspring_metagenome/single_cell_analysis_with_Gabe_Birzu/16S_Read_Alignments/'\n",
    "samples = listdir(filedir+'Alpha_Alleles/paf_files/')\n",
    "\n",
    "fdir = '/Users/harihara/Mount-2/hotspring_metagenome/single_cell_analysis_with_Gabe_Birzu/16S_Gene_Segments/'\n",
    "out = {}\n",
    "for f in ['Synechococcus_Alpha_16SrRNA','Synechococcus_Beta_16SrRNA','Synechococcus_Gamma_16SrRNA']:\n",
    "    lines = open(fdir+f+'.fna').readlines()\n",
    "    s = ''\n",
    "    for l in lines:\n",
    "        if not l.startswith('>'):\n",
    "            s += l.replace(\"\\n\",\"\")\n",
    "    out[f.replace(\"Synechococcus_\",\"\").replace(\"_16SrRNA\",\"\")] = s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_op = pd.DataFrame()\n",
    "Samples = {}\n",
    "\n",
    "for s in samples:\n",
    "    MisMatches = {}\n",
    "    \n",
    "    print(s)\n",
    "    \n",
    "    df_alpha = Load_PAF(filedir+'Alpha_Alleles/paf_files/'+s, s)\n",
    "    df_beta = Load_PAF(filedir+'Beta_Alleles/paf_files/'+s, s)\n",
    "    df_gamma = Load_PAF(filedir+'Gamma_Alleles/paf_files/'+s, s)\n",
    "    \n",
    "    Temp = pd.DataFrame()\n",
    "    Temp = Temp.append(df_alpha).append(df_beta).append(df_gamma)\n",
    "    Temp = Temp[Temp['SEnd'] < 1300]\n",
    "    Temp['Subject'] = Temp['Subject'].str.replace(\"Synechococcus_\",\"\")\n",
    "    Temp['Subject'] = Temp['Subject'].str.replace(\"_16SrRNA\",\"\")\n",
    "    Temp['Divergence'] = Temp['MisMatches']/150*100\n",
    "    \n",
    "    Mismatches = Temp['MisMatches'].unique()\n",
    "    for m in Mismatches:\n",
    "        df_filt = Temp[Temp['MisMatches'] == m]\n",
    "        CS_strings = df_filt['CS'].tolist()\n",
    "        SStarts = df_filt['SStart'].tolist()\n",
    "        SEnds = df_filt['SEnd'].tolist()\n",
    "        subjects = df_filt['Subject'].tolist()\n",
    "    \n",
    "        Alignment_profile = {'Alpha':{'A':np.zeros(1300),'C':np.zeros(1300),\n",
    "                                          'G':np.zeros(1300),'T':np.zeros(1300)},\n",
    "                                 'Beta':{'A':np.zeros(1300),'C':np.zeros(1300),\n",
    "                                         'G':np.zeros(1300),'T':np.zeros(1300)},\n",
    "                                 'Gamma':{'A':np.zeros(1300),'C':np.zeros(1300),\n",
    "                                          'G':np.zeros(1300),'T':np.zeros(1300)}}\n",
    "        \n",
    "        for i in range(0, len(CS_strings)):\n",
    "            out_string = Parse_CS_String(CS_strings[i].replace(\"cs:Z:\",\"\"), \n",
    "                                         out[subjects[i]][SStarts[i]:SEnds[i]])\n",
    "            AP = Generate_Alignment_Profile(out_string)\n",
    "            for n in ['A','C','G','T']:\n",
    "                Alignment_profile[subjects[i]][n][SStarts[i]:SEnds[i]] += AP[n]\n",
    "        MisMatches[m] = Alignment_profile\n",
    "    Samples[s.replace(\"_FD.paf\",\"\")] = MisMatches\n",
    "    df_op = df_op.append(Temp, ignore_index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "mismatches = []\n",
    "for s in Samples:\n",
    "    mis_mat = Samples[s].keys()\n",
    "    mismatches += mis_mat\n",
    "mismatches = list(set(mismatches))\n",
    "divergence = np.array(mismatches)/150*100.0\n",
    "D = np.max(divergence)\n",
    "out_mat = {}\n",
    "for d in np.arange(1, D):\n",
    "    print(d)\n",
    "    x = divergence[divergence < d]\n",
    "    max_mismat = mismatches[:len(x)][-1]\n",
    "    df = pd.DataFrame()\n",
    "    for g in ['Alpha','Beta','Gamma']:\n",
    "        for n in ['A','C','G','T']:\n",
    "            df[(g,n)] = np.zeros(1300)\n",
    "            for s in Samples:\n",
    "                for m in Samples[s]:\n",
    "                    if m < max_mismat: \n",
    "                        df[(g,n)] += Samples[s][m][g][n]\n",
    "    df_normalized = deepcopy(df)\n",
    "\n",
    "    s = (df_normalized[[('Alpha','A'),('Alpha','C'),('Alpha','G'),('Alpha','T')]].sum(axis = 1))\n",
    "    df_normalized['Alpha','A'] = df_normalized[('Alpha','A')]/s\n",
    "    df_normalized['Alpha','C'] = df_normalized[('Alpha','C')]/s\n",
    "    df_normalized['Alpha','G'] = df_normalized[('Alpha','G')]/s\n",
    "    df_normalized['Alpha','T'] = df_normalized[('Alpha','T')]/s\n",
    "\n",
    "    s = (df_normalized[[('Beta','A'),('Beta','C'),('Beta','G'),('Beta','T')]].sum(axis = 1))\n",
    "    df_normalized['Beta','A'] = df_normalized[('Beta','A')]/s\n",
    "    df_normalized['Beta','C'] = df_normalized[('Beta','C')]/s\n",
    "    df_normalized['Beta','G'] = df_normalized[('Beta','G')]/s\n",
    "    df_normalized['Beta','T'] = df_normalized[('Beta','T')]/s\n",
    "\n",
    "    s = (df_normalized[[('Gamma','A'),('Gamma','C'),('Gamma','G'),('Gamma','T')]].sum(axis = 1))\n",
    "    df_normalized['Gamma','A'] = df_normalized[('Gamma','A')]/s\n",
    "    df_normalized['Gamma','C'] = df_normalized[('Gamma','C')]/s\n",
    "    df_normalized['Gamma','G'] = df_normalized[('Gamma','G')]/s\n",
    "    df_normalized['Gamma','T'] = df_normalized[('Gamma','T')]/s\n",
    "    \n",
    "    df_temp = pd.DataFrame()\n",
    "    nucleotides = ['A','C','G', 'T']\n",
    "    for allele in ['Alpha','Beta','Gamma']:\n",
    "        df_temp[allele] = np.zeros(1300)\n",
    "        for i in range(len(nucleotides)):\n",
    "            for j in range(i+1, len(nucleotides)):\n",
    "                df_temp[allele] += (df_normalized[(allele,nucleotides[i])]*df_normalized[(allele,nucleotides[j])])\n",
    "    \n",
    "    out_mat[d] = df_temp.rolling(200).mean()[200:].reset_index()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams.update(rcParams)\n",
    "outpath = '/Users/harihara/Research-Activities/Plots/Hot_Spring_Plots/Single_Cell_Analyses/'\n",
    "\n",
    "fig, ax = plt.subplots(1,1, figsize = (3.43,0.8*3.43), dpi = 80, sharex = True, sharey = True)\n",
    "\n",
    "ax.plot(out_mat[9]['Alpha'].tolist(), color = 'C0', linewidth = 1, label = r'$\\mathcal{H}_{OS\\minus A}$')\n",
    "ax.plot(out_mat[9]['Beta'].tolist(), color = 'C1', linewidth = 1, label = r'$\\mathcal{H}_{OS\\minus B^\\prime}$')\n",
    "ax.plot(out_mat[9]['Gamma'].tolist(), color = 'C2', linewidth = 1, label = r'$\\mathcal{H}_\\gamma$')\n",
    "ax.axvline(400, color = 'black', linestyle = '--')\n",
    "ax.legend(frameon = False, loc = 1, ncol = 1, handlelength = 0.5, bbox_to_anchor=(1.05,1.05))\n",
    "ax.set_ylabel(r'average diversity index, $\\mathcal{H}$')\n",
    "fig.tight_layout()\n",
    "fig.savefig(outpath+'Diversity-Index.pdf')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
