{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from os import listdir\n",
    "from copy import deepcopy\n",
    "from Utils_BLAST import *\n",
    "\n",
    "filedir = '/Users/harihara/Research-Activities/Data/Hot-Spring/Single_Cell_Gabe/BLAST_Hits/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A\n",
      "A YSG_1514\n",
      "A YSG_0675\n",
      "A YSG_0096b\n",
      "A YSG_1709\n",
      "A YSG_1325\n",
      "A YSG_0764\n",
      "A YSG_1272\n",
      "A YSG_0688\n",
      "A YSG_1694\n",
      "A YSG_1779\n",
      "A YSG_0072a\n",
      "A YSG_0548\n",
      "A YSG_1367\n",
      "A YSG_0694b\n",
      "A YSG_1299\n",
      "A YSG_0966\n",
      "A YSG_1319\n",
      "A YSG_1924\n",
      "A YSG_0108a\n",
      "A YSG_0136a\n",
      "A YSG_0168b\n",
      "A YSG_1576\n",
      "A YSG_0898\n",
      "A YSG_0340\n",
      "A YSG_0314\n",
      "A YSG_0985\n",
      "A YSG_1142\n",
      "A YSG_1748\n",
      "A YSG_1447b\n",
      "A YSG_1487\n",
      "A YSG_0699\n",
      "B\n",
      "B YSG_1779\n",
      "B YSG_0985\n",
      "B YSG_1514\n",
      "B YSG_0340\n",
      "B YSG_1694\n",
      "B YSG_0548\n",
      "B YSG_1367\n",
      "B YSG_0675\n",
      "B YSG_0694b\n",
      "B YSG_0764\n",
      "B YSG_1272\n",
      "B YSG_0108a\n",
      "B YSG_0096b\n",
      "B YSG_1142\n",
      "B YSG_0136a\n",
      "B YSG_0688\n",
      "B YSG_1748\n",
      "B YSG_0072a\n",
      "B YSG_0314\n",
      "B YSG_0966\n",
      "B YSG_0168b\n",
      "B YSG_1299\n",
      "B YSG_0898\n",
      "B YSG_1325\n",
      "B YSG_1924\n",
      "B YSG_1487\n",
      "B YSG_1319\n",
      "B YSG_1576\n",
      "B YSG_1709\n",
      "B YSG_1447b\n",
      "B YSG_0699\n",
      "C\n",
      "C YSG_0096b\n",
      "C YSG_0072a\n",
      "C YSG_1142\n",
      "C YSG_0966\n",
      "C YSG_1325\n",
      "C YSG_1299\n",
      "C YSG_1924\n",
      "C YSG_0699\n",
      "C YSG_0136a\n",
      "C YSG_0340\n",
      "C YSG_1779\n",
      "C YSG_0314\n",
      "C YSG_0168b\n",
      "C YSG_0548\n",
      "C YSG_0688\n",
      "C YSG_1367\n",
      "C YSG_0675\n",
      "C YSG_1272\n",
      "C YSG_1447b\n",
      "C YSG_1514\n",
      "C YSG_1709\n",
      "C YSG_1748\n",
      "C YSG_0985\n",
      "C YSG_1694\n",
      "C YSG_0898\n",
      "C YSG_1319\n",
      "C YSG_1487\n",
      "C YSG_1576\n",
      "C YSG_0108a\n",
      "C YSG_0764\n",
      "C YSG_0694b\n"
     ]
    }
   ],
   "source": [
    "df_Gene_Blast_ = pd.DataFrame()\n",
    "for Allele in ['A','B','C']:\n",
    "    print(Allele)\n",
    "    files = listdir(filedir+Allele+'_Alleles/')\n",
    "    for g in files:\n",
    "        if g.startswith('YSG'):\n",
    "            splits = g.split('_')\n",
    "            gene = splits[0]+'_'+splits[1]\n",
    "            samples = listdir(filedir+Allele+'_Alleles/'+g+'/')\n",
    "            for s in samples:\n",
    "                if s.startswith('MS'):\n",
    "                    blast_hits = listdir(filedir+Allele+'_Alleles/'+g+'/'+s+'/')\n",
    "                    for b in blast_hits:\n",
    "                        if b.startswith('MS'):\n",
    "                            df = Load_BLAST_Hits(filedir+Allele+'_Alleles/'+g+'/'+s+'/'+b)\n",
    "                            df['Sample'] = s\n",
    "                            df['Allele'] = Allele\n",
    "                            df['Gene'] = gene\n",
    "                            df_Gene_Blast_ = df_Gene_Blast_.append(df, ignore_index = True)\n",
    "            print(Allele, gene)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50 3455242\n",
      "All Reads\n",
      "70 3335339\n",
      "All Reads\n",
      "75 3298130\n",
      "All Reads\n",
      "80 3264169\n",
      "All Reads\n",
      "85 3215413\n",
      "All Reads\n",
      "90 3157503\n",
      "All Reads\n",
      "95 3022464\n",
      "All Reads\n",
      "97 2885078\n",
      "All Reads\n",
      "100 2309737\n",
      "All Reads\n"
     ]
    }
   ],
   "source": [
    "cov_filter = [50, 70, 75, 80, 85, 90, 95, 97, 100]\n",
    "df_sample_wise = pd.DataFrame()\n",
    "df_gene_wise = pd.DataFrame()\n",
    "\n",
    "for c in cov_filter:\n",
    "    df_Gene_Blast = df_Gene_Blast_[df_Gene_Blast_['qcov']>=c]\n",
    "    idx = df_Gene_Blast.groupby(['Sample','Allele',\n",
    "                                 'qseqid'])['bitscore'].transform(max) == df_Gene_Blast['bitscore']\n",
    "    df_Gene_Blast = df_Gene_Blast.loc[idx]\n",
    "    df_Gene_Blast_grp = pd.DataFrame()\n",
    "    df_Gene_Blast_grp['StrainList'] = df_Gene_Blast.groupby(['Gene', 'Sample', 'qseqid'])['Allele'].apply(list)\n",
    "    df_Gene_Blast_grp['Strain_String'] = df_Gene_Blast_grp['StrainList'].apply(To_String)\n",
    "    df_Gene_Blast = df_Gene_Blast.set_index(['Gene','Sample','qseqid'])\n",
    "    df_Gene_Blast = df_Gene_Blast.join(df_Gene_Blast_grp[['Strain_String']])\n",
    "    df_Gene_Blast = df_Gene_Blast.reset_index()\n",
    "    df_Gene_Blast['Cov'] = c\n",
    "    #BLAST_Hits_Top_30_Alleles_1E-3_qcov_'+str(c)+'.pkl'\n",
    "    df_Gene_Blast.to_pickle(filedir+'BLAST_Hits_Top_30_Alleles_qcov_'+str(c)+'.pkl')\n",
    "    print(c, len(df_Gene_Blast))\n",
    "    \n",
    "    gene_wise_ABC, sample_wise_ABC = Pick_Minimum_SNP(df_Gene_Blast)\n",
    "    gene_wise_ABC['Type'] = 'ABC'\n",
    "    gene_wise_ABC['Cov'] = c\n",
    "    sample_wise_ABC['Type'] = 'ABC'\n",
    "    sample_wise_ABC['Cov'] = c\n",
    "    df_gene_wise = df_gene_wise.append(gene_wise_ABC.reset_index(), ignore_index = True)\n",
    "    df_sample_wise = df_sample_wise.append(sample_wise_ABC.reset_index(), ignore_index = True)\n",
    "\n",
    "    gene_wise_AB, sample_wise_AB = Pick_Minimum_SNP(df_Gene_Blast[df_Gene_Blast['Allele'] != 'C'])\n",
    "    gene_wise_AB['Type'] = 'AB'\n",
    "    gene_wise_AB['Cov'] = c\n",
    "    sample_wise_AB['Type'] = 'AB'\n",
    "    sample_wise_AB['Cov'] = c\n",
    "    df_gene_wise = df_gene_wise.append(gene_wise_AB.reset_index(), ignore_index = True)\n",
    "    df_sample_wise = df_sample_wise.append(sample_wise_AB.reset_index(), ignore_index = True)\n",
    "\n",
    "    print('All Reads')\n",
    "    (gene_wise_all_ABC, \n",
    "     sample_wise_all_ABC) = Pick_Minimum_SNP(df_Gene_Blast[df_Gene_Blast['Strain_String'] == 'ABC'])\n",
    "    gene_wise_all_ABC['Type'] = 'All_ABC'\n",
    "    gene_wise_all_ABC['Cov'] = c\n",
    "    sample_wise_all_ABC['Type'] = 'All_ABC'\n",
    "    sample_wise_all_ABC['Cov'] = c\n",
    "    df_gene_wise = df_gene_wise.append(gene_wise_all_ABC.reset_index(), ignore_index = True)\n",
    "    df_sample_wise = df_sample_wise.append(sample_wise_all_ABC.reset_index(), ignore_index = True)\n",
    "\n",
    "    (gene_wise_all_AB, \n",
    "     sample_wise_all_AB) = Pick_Minimum_SNP(df_Gene_Blast[(df_Gene_Blast['Strain_String'] == 'ABC') & \n",
    "                                                          (df_Gene_Blast['Allele'] != 'C')])\n",
    "    gene_wise_all_AB['Type'] = 'All_AB'\n",
    "    gene_wise_all_AB['Cov'] = c\n",
    "    sample_wise_all_AB['Type'] = 'All_AB'\n",
    "    sample_wise_all_AB['Cov'] = c\n",
    "    df_gene_wise = df_gene_wise.append(gene_wise_all_AB.reset_index(), ignore_index = True)\n",
    "    df_sample_wise = df_sample_wise.append(sample_wise_all_AB.reset_index(), ignore_index = True)\n",
    "\n",
    "df_gene_wise.to_pickle(filedir+'BLAST_Hits_Top_30_Alleles_Gene_Wise.pkl')\n",
    "df_sample_wise.to_pickle(filedir+'BLAST_Hits_Top_30_Alleles_Sample_Wise.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '/Users/harihara/Mount-2/hotspring_metagenome/single_cell_analysis_with_Gabe_Birzu/16S_Blast_Alignments/'\n",
    "allele_types = ['Alpha','Beta','Gamma']\n",
    "Alleles = ['A','B','C']\n",
    "df_Alleles = pd.DataFrame()\n",
    "\n",
    "for i in range(0, len(allele_types)):\n",
    "    Allele = Alleles[i]\n",
    "    samples = listdir(data_dir+allele_types[i]+'_Alleles/paf_files/')\n",
    "    for s in samples:\n",
    "        if s.startswith('HotsprSample'):\n",
    "            sample = s.replace(\"HotsprSample\",\"\").replace(\"_FD.paf\",\"\")\n",
    "            print(sample)\n",
    "            df = Load_PAF(data_dir+allele_types[i]+'_Alleles/paf_files/'+s, sample, flip = True)\n",
    "            df['Allele'] = Allele\n",
    "            df_Alleles = df_Alleles.append(df)\n",
    "df_Alleles['QCov'] = df_Alleles['AlignLength']/150.0*100\n",
    "df_Alleles = df_Alleles[df_Alleles['QCov'] >= 95]\n",
    "df_Alleles['Divergence'] = df_Alleles['MisMatches']/df_Alleles['QLen']*100.0\n",
    "df_Alleles = df_Alleles[['Query', 'Subject','Allele','MisMatches','Divergence','Sample']]\n",
    "df_Alleles = df_Alleles.rename(columns = {'Query':'qseqid', 'Subject':'Gene', 'MisMatches':'mismatch'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_16S_Blast_grp = pd.DataFrame()\n",
    "df_16S_Blast_grp['StrainList'] = df_Alleles.groupby(['Sample', 'qseqid'])['Allele'].apply(list)\n",
    "df_16S_Blast_grp['Strain_String'] = df_16S_Blast_grp['StrainList'].apply(To_String)\n",
    "df_Alleles = df_Alleles.set_index(['Sample','qseqid'])\n",
    "df_Alleles = df_Alleles.join(df_16S_Blast_grp[['Strain_String']])\n",
    "df_Alleles = df_Alleles.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_Alleles.to_pickle(filedir+'BLAST_Hits_16S_rRNA.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
