{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.gridspec import GridSpec\n",
    "from matplotlib.patches import Rectangle\n",
    "\n",
    "import resource\n",
    "soft, hard = resource.getrlimit(resource.RLIMIT_NOFILE)\n",
    "resource.setrlimit(resource.RLIMIT_NOFILE, (hard, hard))\n",
    "\n",
    "from os import listdir\n",
    "from PyPDF2 import PdfMerger\n",
    "from os.path import isdir, isfile\n",
    "from PyCircos_Plots.PyCircos_Utils import *\n",
    "\n",
    "rcParams = {'font.size': 24, 'font.weight': 'normal', 'font.family': 'sans-serif',\n",
    "            'axes.unicode_minus':False, 'axes.labelweight':'normal'}\n",
    "\n",
    "plt.rcParams.update(rcParams)\n",
    "\n",
    "osa_len = 2932766\n",
    "osb_len = 3046682\n",
    "\n",
    "def Make_Circle_Plots(ax, values, color, text, ylim):\n",
    "    angles = np.linspace(0, 2*np.pi, len(values), endpoint=False)\n",
    "    ax.plot(angles, values, linewidth=5, color = color)\n",
    "    ax.grid(False)\n",
    "    ax.set_ylim([0, ylim])\n",
    "    ax.set_yticklabels([])\n",
    "    ax.set_xticks([])\n",
    "    ax.text(0,0, text, horizontalalignment='center', verticalalignment='center',)\n",
    "    ax.spines['polar'].set_visible(False)\n",
    "\n",
    "def Load_Prodigal_GBFF(filepath):\n",
    "    lines = open(filepath).readlines()\n",
    "    op = []\n",
    "    ctr = 1\n",
    "    for i in range(len(lines)):\n",
    "        l = lines[i]\n",
    "        l = l.replace(\"\\n\",\"\")\n",
    "        l = l.strip()\n",
    "        if l == \"//\":\n",
    "            ctr = 1\n",
    "        else:\n",
    "            if l.startswith(\"DEFINITION\"):\n",
    "                l = l.replace(\"DEFINITION \",\"\")\n",
    "                splits = l.split(\";\")\n",
    "                for s in splits:\n",
    "                    if s.startswith(\"seqhdr\"):\n",
    "                        sequence_id = s.replace(\"seqhdr=\",\"\").replace(\"\\\"\",\"\")\n",
    "                    if s.startswith(\"seqlen\"):\n",
    "                        seqlen = int(s.replace(\"seqlen=\",\"\").replace(\"\\\"\",\"\"))\n",
    "            if l.startswith(\"CDS\"):\n",
    "                l = l.replace(\"CDS\",\"\").strip()\n",
    "                if l.startswith(\"complement\"):\n",
    "                    orientation = '-'\n",
    "                else:\n",
    "                    orientation = '+'\n",
    "                l = l.replace(\"complement\",\"\").replace(\"(\",\"\").replace(\")\",\"\").replace(\"<\",\"\").replace(\">\",\"\")\n",
    "                start,end = l.split(\"..\")\n",
    "                i += 1\n",
    "                record = lines[i]\n",
    "                record = record.strip()\n",
    "                if record.startswith(\"/note\"):\n",
    "                    record = record.replace(\"/note=\",\"\").replace(\"\\\"\",\"\")\n",
    "                    splits = record.split(\";\")\n",
    "                    d = {'Query':sequence_id, 'Pred':sequence_id+'_'+str(ctr), 'Qlen':seqlen, \n",
    "                         'Orientation':orientation, 'Start':start, 'End':end}\n",
    "                    for s in splits[:-1]:\n",
    "                        s = s.replace(\"\\n\",\"\")\n",
    "                        key, value = s.split('=')\n",
    "                        d[key] = value\n",
    "                    ctr += 1\n",
    "                    op.append(d)\n",
    "    df_Prodigal_Hits = pd.DataFrame(op)\n",
    "    return df_Prodigal_Hits[['Query','Pred','Orientation','Start','End','Qlen','ID','partial','conf']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "novel_contigs = {}\n",
    "novel_contig_path = '/Users/harihara/Mount-2/hotspring_metagenome/Synechococcus_paper_analysis/\\\n",
    "Ref_Guided_Scaffolding_Clustering_Aug_2022/Missing_Contig_Coords_Aug/'\n",
    "samples = listdir(novel_contig_path+'OSA/')\n",
    "for s in samples:\n",
    "    if(s.startswith(\"Hot\") and s.endswith(\".txt\")):\n",
    "        df_osa = pd.read_csv(novel_contig_path+'OSA/'+s, sep = \"\\t\")\n",
    "        df_osa.loc[(df_osa['Start'] < 0), 'Start'] += osa_len\n",
    "        df_osa.loc[(df_osa['End'] < 0), 'End'] += osa_len\n",
    "        df_osa_grp = df_osa.sort_values(by = ['Contig','Start']).groupby(['Contig']).apply(Max_Clique_Interval_Graph)\n",
    "    \n",
    "        df_osb = pd.read_csv(novel_contig_path+'OSB/'+s, sep = \"\\t\")\n",
    "        df_osb.loc[(df_osb['Start'] < 0), 'Start'] += osb_len\n",
    "        df_osb.loc[(df_osb['End'] < 0), 'End'] += osb_len\n",
    "        df_osb_grp = df_osb.sort_values(by = ['Contig','Start']).groupby(['Contig']).apply(Max_Clique_Interval_Graph)\n",
    "    \n",
    "        novel_contigs[s.replace(\".txt\",\"\")] = {'OSA':df_osa_grp,'OSB':df_osb_grp}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "grp_path = '/Users/harihara/Mount-2/hotspring_metagenome/Synechococcus_paper_analysis/\\\n",
    "Ref_Guided_Scaffolding_Clustering_Aug_2022/BLAST_All_vs_All/Containment_Clusters_Filtered_90_75_1000.txt'\n",
    "df_novel_filtered = pd.read_csv(grp_path, sep = \"\\t\")\n",
    "d = df_novel_filtered.groupby('GroupID')['Contig'].apply(list).to_dict()\n",
    "d_representatives = df_novel_filtered.groupby('GroupID')['RepresentativeContig'].apply(list).to_dict()\n",
    "\n",
    "out_dir = '/Users/harihara/Research-Activities/Plots/Hot_Spring_Plots/Synechococcus-Paper/Novel_Groups_Circos_Aug_2022/'\n",
    "if not isdir(out_dir):\n",
    "    mkdir(out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "prodigal_out = '/Users/harihara/Mount-2/hotspring_metagenome/Synechococcus_paper_analysis/\\\n",
    "Ref_Guided_Scaffolding_Clustering_Aug_2022/BLAST_All_vs_All/Prodigal/Representatives_Prodigal.out'\n",
    "df_prodigal_hits = Load_Prodigal_GBFF(prodigal_out)\n",
    "df_prodigal_hits = df_prodigal_hits.set_index('Pred')\n",
    "\n",
    "eggnog_path = '/Users/harihara/Mount-2/hotspring_metagenome/Synechococcus_paper_analysis/\\\n",
    "Ref_Guided_Scaffolding_Clustering_Aug_2022/BLAST_All_vs_All/EggNOG/Representatives.eggnog.out.emapper.annotations'\n",
    "df_eggnog = pd.read_csv(eggnog_path, sep = \"\\t\")\n",
    "df_eggnog = df_eggnog[['#query','seed_ortholog','evalue','score','max_annot_lvl',\n",
    "                       'COG_category','Description','Preferred_name']]\n",
    "df_eggnog = df_eggnog.rename(columns = {'#query':'Query'})\n",
    "df_eggnog = df_eggnog.set_index('Query')\n",
    "\n",
    "df_prodigal_hits = df_prodigal_hits.join(df_eggnog)\n",
    "reps = np.unique(df_novel_filtered['RepresentativeContig'].tolist())\n",
    "\n",
    "df_prodigal_hits = df_prodigal_hits.reset_index()\n",
    "df_prodigal_hits = df_prodigal_hits.merge(df_novel_filtered[['RepresentativeContig','GroupID']].drop_duplicates(),\n",
    "                                          left_on = 'Query', right_on='RepresentativeContig', how = 'left')\n",
    "df_prodigal_hits['COG_category'] = df_prodigal_hits['COG_category'].fillna(\"-\")\n",
    "\n",
    "colors = {'C' : 'red', 'L':'blue', 'P':'green', '-':'grey', 'M':'orange', 'I':'gold', \n",
    "          'S':'purple', 'H':'olive', 'J':'yellow', 'E':'cyan', 'G':'magenta', 'V':'lime',\n",
    "          'O':'teal', 'F':'black', 'Q':'yellow', 'T':'lightcoral', 'KL':'brown', 'K':'peachpuff', 'FP':'plum', \n",
    "          'MU':'peru', 'KLT':'sienna', 'KT':'maroon', 'NU':'indigo', 'U':'darkolivegreen',\n",
    "          'GM':'crimson', 'NOU':'thistle', 'NPTU':'dodgerblue', 'EGP':'firebrick', 'IQ':'mediumslateblue', \n",
    "          'PT':'lightsalmon', 'D':'khaki', 'OU':'darkseagreen', 'ET':'lawngreen', 'CO':'sandybrown',\n",
    "          'CH':'tan', 'UW':'royalblue', 'HP':'goldenrod', 'EQ':'springgreen', 'HJ':'cadetblue', 'CDZ':'pink', \n",
    "          'FG':'slategrey'}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "synechococcus_blast = '/Users/harihara/Mount-2/hotspring_metagenome/Synechococcus_paper_analysis/\\\n",
    "Ref_Guided_Scaffolding_Clustering_Aug_2022/BLAST_All_vs_All/Representatives.Syenchococcus.blast'\n",
    "\n",
    "df_blast = pd.read_csv(synechococcus_blast, sep = \"\\t\",names=['qseqid', 'sseqid', 'pident', 'length', 'mismatch', \n",
    "                                                              'gapopen', 'qlen', 'qstart', 'qend', 'slen', \n",
    "                                                              'sstart', 'send', 'evalue', 'bitscore'])\n",
    "df_blast = df_blast.merge(df_novel_filtered[['RepresentativeContig','GroupID']].drop_duplicates(),\n",
    "                          left_on = 'qseqid', right_on='RepresentativeContig', how = 'left')\n",
    "df_blast['sseqid'] = df_blast['sseqid'].replace(\"gi|86604733|ref|NC_007775.1|\",\"OSA\")\n",
    "df_blast['sseqid'] = df_blast['sseqid'].replace(\"gi|86607503|ref|NC_007776.1|\",\"OSB\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Group_595\n",
      "Group_603\n",
      "Group_604\n",
      "Group_605\n",
      "Group_609\n",
      "Group_612\n",
      "Group_614\n",
      "Group_621\n",
      "Group_622\n",
      "Group_631\n",
      "Group_639\n",
      "Group_64\n",
      "Group_645\n",
      "Group_648\n",
      "Group_651\n",
      "Group_659\n",
      "Group_662\n",
      "Group_67\n",
      "Group_699\n",
      "Group_704\n",
      "Group_71\n",
      "Group_718\n",
      "Group_720\n",
      "Group_721\n",
      "Group_73\n"
     ]
    }
   ],
   "source": [
    "plt.rcParams.update(rcParams)\n",
    "\n",
    "for g in list(d.keys())[200:225]:    \n",
    "    print(g)\n",
    "    contigs = d[g]+[d_representatives[g][0]]\n",
    "    osa_contig_count, osb_contig_count = 0, 0\n",
    "    osa_contigs, osb_contigs = [], []\n",
    "    df_osa, df_osb = pd.DataFrame(), pd.DataFrame()\n",
    "    \n",
    "    for c in contigs:\n",
    "        splits = c.split('_')\n",
    "        contig = splits[-2]+'_'+splits[-1]\n",
    "        genome = splits[-3].upper()\n",
    "        sample = \"_\".join(splits[:-3])\n",
    "        \n",
    "        if genome == \"OSA\": osa_contig_count += 1\n",
    "        if genome == \"OSB\": osb_contig_count += 1\n",
    "        \n",
    "        try:\n",
    "            row = novel_contigs[sample][genome].loc[contig]\n",
    "            row['Group'] = g\n",
    "            row['Sample'] = sample\n",
    "            if genome == 'OSA': df_osa = df_osa.append(row)\n",
    "            elif genome == 'OSB': df_osb = df_osb.append(row)\n",
    "        except:\n",
    "            print(\"Missing...\",sample,genome,contig)\n",
    "            pass\n",
    "    \n",
    "    if len(df_osa) > 0: df_osa = df_osa.reset_index()\n",
    "    if len(df_osb) > 0: df_osb = df_osb.reset_index()\n",
    "    \n",
    "    osa_counts, osb_counts, osa_gbl_counts, osb_gbl_counts = Make_Counts(df_osa, df_osb, osa_len, osb_len)\n",
    "    osa_gbl_counts += 25\n",
    "    osb_gbl_counts += 25\n",
    "    \n",
    "    ylim = max(np.max(osa_gbl_counts), np.max(osb_gbl_counts))\n",
    "    \n",
    "    fig = plt.figure(figsize=(20, 15))\n",
    "    gs = GridSpec(nrows=2, ncols=2, height_ratios=[2.5, 1.5])\n",
    "    ax0 = fig.add_subplot(gs[0, 0], projection = 'polar')\n",
    "    ax1 = fig.add_subplot(gs[0, 1], projection = 'polar')\n",
    "    ax2 = fig.add_subplot(gs[1, :])\n",
    "    \n",
    "    starts = df_prodigal_hits.loc[df_prodigal_hits['GroupID'] == g, 'Start'].tolist()\n",
    "    ends = df_prodigal_hits.loc[df_prodigal_hits['GroupID'] == g, 'End'].tolist()\n",
    "    orientations = df_prodigal_hits.loc[df_prodigal_hits['GroupID'] == g, 'Orientation'].tolist()\n",
    "    cog_categories = df_prodigal_hits.loc[df_prodigal_hits['GroupID'] == g, 'COG_category'].tolist()\n",
    "\n",
    "    osa_alignments_starts = df_blast.loc[(df_blast['GroupID'] == g) & (df_blast['sseqid'] == 'OSA') ,\n",
    "                                         'qstart'].tolist()\n",
    "    osa_alignments_ends = df_blast.loc[(df_blast['GroupID'] == g) & (df_blast['sseqid'] == 'OSA') ,\n",
    "                                         'qend'].tolist()\n",
    "    \n",
    "    osb_alignments_starts = df_blast.loc[(df_blast['GroupID'] == g) & (df_blast['sseqid'] == 'OSB') ,\n",
    "                                         'qstart'].tolist()\n",
    "    osb_alignments_ends = df_blast.loc[(df_blast['GroupID'] == g) & (df_blast['sseqid'] == 'OSB') ,\n",
    "                                         'qend'].tolist()\n",
    "    \n",
    "    width = df_prodigal_hits.loc[df_prodigal_hits['GroupID'] == g, 'Qlen'].tolist()[0]\n",
    "    ax2.plot([4, width],[4,4], linewidth=20, color = 'black')\n",
    "    \n",
    "    for i in range(len(osa_alignments_starts)):\n",
    "        start = min(osa_alignments_starts[i], osa_alignments_ends[i])\n",
    "        end = max(osa_alignments_starts[i], osa_alignments_ends[i])\n",
    "        ax2.plot([start, end], [3, 3], color = 'red', linewidth = 2)\n",
    "        \n",
    "    for i in range(len(osb_alignments_starts)):\n",
    "        start = min(osb_alignments_starts[i], osb_alignments_ends[i])\n",
    "        end = max(osb_alignments_starts[i], osb_alignments_ends[i])\n",
    "        ax2.plot([start, end], [2, 2], color = 'blue', linewidth = 2)\n",
    "    \n",
    "    for i in range(0, len(starts)):\n",
    "        if orientations[i] == '+':\n",
    "            ax2.arrow(int(starts[i]),1,int(ends[i])-int(starts[i])+1,0, width = 0.25, alpha = 0.8, \n",
    "                      length_includes_head = True, head_length = 20, color = colors[cog_categories[i]])\n",
    "        else:\n",
    "            ax2.arrow(int(ends[i]),1,int(starts[i])-int(ends[i])+1,0, width = 0.25, alpha = 0.8, \n",
    "                      length_includes_head = True, head_length = 20, color = colors[cog_categories[i]])\n",
    "    ax2.set_ylim([-1, 5])\n",
    "    ax2.set_yticks([])\n",
    "    fig.tight_layout()\n",
    "    \n",
    "    label=(\"Synechococcus Sub.Sp A\\n\"+\"#Contigs (in Group):\"+str(osa_contig_count))\n",
    "    Make_Circle_Plots(ax0, osa_gbl_counts, 'green', label, ylim)\n",
    "    label=(\"Synechococcus Sub.Sp B\\n\"+\"#Contigs (in Group):\"+str(osb_contig_count))\n",
    "    Make_Circle_Plots(ax1, osb_gbl_counts, 'orange', label, ylim)\n",
    "    fig.suptitle(g)\n",
    "    fig.tight_layout()\n",
    "    fig.savefig(out_dir+g+'.pdf')\n",
    "    \n",
    "    plt.close(\"all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Group_0.pdf\n",
      "Group_1.pdf\n",
      "Group_10.pdf\n",
      "Group_100.pdf\n",
      "Group_101.pdf\n",
      "Group_108.pdf\n",
      "Group_110.pdf\n",
      "Group_112.pdf\n",
      "Group_115.pdf\n",
      "Group_118.pdf\n",
      "Group_119.pdf\n",
      "Group_120.pdf\n",
      "Group_121.pdf\n",
      "Group_122.pdf\n",
      "Group_123.pdf\n",
      "Group_127.pdf\n",
      "Group_129.pdf\n",
      "Group_130.pdf\n",
      "Group_131.pdf\n",
      "Group_132.pdf\n",
      "Group_135.pdf\n",
      "Group_138.pdf\n",
      "Group_139.pdf\n",
      "Group_14.pdf\n",
      "Group_140.pdf\n",
      "Group_15.pdf\n",
      "Group_152.pdf\n",
      "Group_155.pdf\n",
      "Group_160.pdf\n",
      "Group_162.pdf\n",
      "Group_17.pdf\n",
      "Group_173.pdf\n",
      "Group_174.pdf\n",
      "Group_178.pdf\n",
      "Group_179.pdf\n",
      "Group_18.pdf\n",
      "Group_182.pdf\n",
      "Group_189.pdf\n",
      "Group_191.pdf\n",
      "Group_193.pdf\n",
      "Group_197.pdf\n",
      "Group_198.pdf\n",
      "Group_199.pdf\n",
      "Group_201.pdf\n",
      "Group_204.pdf\n",
      "Group_205.pdf\n",
      "Group_208.pdf\n",
      "Group_21.pdf\n",
      "Group_212.pdf\n",
      "Group_222.pdf\n",
      "Group_223.pdf\n",
      "Group_226.pdf\n",
      "Group_227.pdf\n",
      "Group_229.pdf\n",
      "Group_235.pdf\n",
      "Group_237.pdf\n",
      "Group_244.pdf\n",
      "Group_249.pdf\n",
      "Group_254.pdf\n",
      "Group_256.pdf\n",
      "Group_257.pdf\n",
      "Group_258.pdf\n",
      "Group_259.pdf\n",
      "Group_266.pdf\n",
      "Group_267.pdf\n",
      "Group_268.pdf\n",
      "Group_269.pdf\n",
      "Group_27.pdf\n",
      "Group_270.pdf\n",
      "Group_273.pdf\n",
      "Group_279.pdf\n",
      "Group_282.pdf\n",
      "Group_284.pdf\n",
      "Group_287.pdf\n",
      "Group_29.pdf\n",
      "Group_291.pdf\n",
      "Group_292.pdf\n",
      "Group_293.pdf\n",
      "Group_294.pdf\n",
      "Group_298.pdf\n",
      "Group_30.pdf\n",
      "Group_306.pdf\n",
      "Group_307.pdf\n",
      "Group_308.pdf\n",
      "Group_309.pdf\n",
      "Group_326.pdf\n",
      "Group_327.pdf\n",
      "Group_33.pdf\n",
      "Group_332.pdf\n",
      "Group_333.pdf\n",
      "Group_334.pdf\n",
      "Group_338.pdf\n",
      "Group_339.pdf\n",
      "Group_343.pdf\n",
      "Group_344.pdf\n",
      "Group_35.pdf\n",
      "Group_351.pdf\n",
      "Group_352.pdf\n",
      "Group_353.pdf\n",
      "Group_356.pdf\n",
      "Group_358.pdf\n",
      "Group_359.pdf\n",
      "Group_36.pdf\n",
      "Group_360.pdf\n",
      "Group_365.pdf\n",
      "Group_366.pdf\n",
      "Group_367.pdf\n",
      "Group_37.pdf\n",
      "Group_371.pdf\n",
      "Group_373.pdf\n",
      "Group_374.pdf\n",
      "Group_377.pdf\n",
      "Group_378.pdf\n",
      "Group_38.pdf\n",
      "Group_380.pdf\n",
      "Group_381.pdf\n",
      "Group_387.pdf\n",
      "Group_39.pdf\n",
      "Group_390.pdf\n",
      "Group_393.pdf\n",
      "Group_394.pdf\n",
      "Group_395.pdf\n",
      "Group_4.pdf\n",
      "Group_409.pdf\n",
      "Group_410.pdf\n",
      "Group_411.pdf\n",
      "Group_412.pdf\n",
      "Group_414.pdf\n",
      "Group_418.pdf\n",
      "Group_419.pdf\n",
      "Group_42.pdf\n",
      "Group_422.pdf\n",
      "Group_43.pdf\n",
      "Group_430.pdf\n",
      "Group_435.pdf\n",
      "Group_438.pdf\n",
      "Group_44.pdf\n",
      "Group_442.pdf\n",
      "Group_444.pdf\n",
      "Group_447.pdf\n",
      "Group_453.pdf\n",
      "Group_454.pdf\n",
      "Group_455.pdf\n",
      "Group_456.pdf\n",
      "Group_457.pdf\n",
      "Group_459.pdf\n",
      "Group_463.pdf\n",
      "Group_468.pdf\n",
      "Group_469.pdf\n",
      "Group_470.pdf\n",
      "Group_474.pdf\n",
      "Group_477.pdf\n",
      "Group_478.pdf\n",
      "Group_479.pdf\n",
      "Group_480.pdf\n",
      "Group_482.pdf\n",
      "Group_483.pdf\n",
      "Group_485.pdf\n",
      "Group_487.pdf\n",
      "Group_489.pdf\n",
      "Group_495.pdf\n",
      "Group_497.pdf\n",
      "Group_498.pdf\n",
      "Group_500.pdf\n",
      "Group_502.pdf\n",
      "Group_503.pdf\n",
      "Group_504.pdf\n",
      "Group_505.pdf\n",
      "Group_506.pdf\n",
      "Group_513.pdf\n",
      "Group_520.pdf\n",
      "Group_521.pdf\n",
      "Group_522.pdf\n",
      "Group_523.pdf\n",
      "Group_525.pdf\n",
      "Group_53.pdf\n",
      "Group_534.pdf\n",
      "Group_535.pdf\n",
      "Group_536.pdf\n",
      "Group_54.pdf\n",
      "Group_540.pdf\n",
      "Group_551.pdf\n",
      "Group_552.pdf\n",
      "Group_553.pdf\n",
      "Group_555.pdf\n",
      "Group_559.pdf\n",
      "Group_560.pdf\n",
      "Group_563.pdf\n",
      "Group_564.pdf\n",
      "Group_565.pdf\n",
      "Group_566.pdf\n",
      "Group_569.pdf\n",
      "Group_57.pdf\n",
      "Group_578.pdf\n",
      "Group_579.pdf\n",
      "Group_58.pdf\n",
      "Group_581.pdf\n",
      "Group_583.pdf\n",
      "Group_592.pdf\n",
      "Group_594.pdf\n",
      "Group_595.pdf\n",
      "Group_603.pdf\n",
      "Group_604.pdf\n",
      "Group_605.pdf\n",
      "Group_609.pdf\n",
      "Group_612.pdf\n",
      "Group_614.pdf\n",
      "Group_621.pdf\n",
      "Group_622.pdf\n",
      "Group_631.pdf\n",
      "Group_639.pdf\n",
      "Group_64.pdf\n",
      "Group_645.pdf\n",
      "Group_648.pdf\n",
      "Group_651.pdf\n",
      "Group_659.pdf\n",
      "Group_662.pdf\n",
      "Group_67.pdf\n",
      "Group_699.pdf\n",
      "Group_704.pdf\n",
      "Group_71.pdf\n",
      "Group_718.pdf\n",
      "Group_720.pdf\n",
      "Group_721.pdf\n",
      "Group_73.pdf\n",
      "Group_733.pdf\n",
      "Group_735.pdf\n",
      "Group_736.pdf\n",
      "Group_737.pdf\n",
      "Group_74.pdf\n",
      "Group_749.pdf\n",
      "Group_75.pdf\n",
      "Group_754.pdf\n",
      "Group_76.pdf\n",
      "Group_762.pdf\n",
      "Group_763.pdf\n",
      "Group_772.pdf\n",
      "Group_78.pdf\n",
      "Group_788.pdf\n",
      "Group_79.pdf\n",
      "Group_796.pdf\n",
      "Group_8.pdf\n",
      "Group_803.pdf\n",
      "Group_806.pdf\n",
      "Group_807.pdf\n",
      "Group_809.pdf\n",
      "Group_81.pdf\n",
      "Group_813.pdf\n",
      "Group_816.pdf\n",
      "Group_825.pdf\n",
      "Group_836.pdf\n",
      "Group_842.pdf\n",
      "Group_844.pdf\n",
      "Group_85.pdf\n",
      "Group_856.pdf\n",
      "Group_880.pdf\n",
      "Group_885.pdf\n",
      "Group_89.pdf\n",
      "Group_897.pdf\n",
      "Group_90.pdf\n",
      "Group_902.pdf\n",
      "Group_906.pdf\n",
      "Group_908.pdf\n",
      "Group_91.pdf\n",
      "Group_912.pdf\n",
      "Group_914.pdf\n",
      "Group_92.pdf\n",
      "Group_921.pdf\n",
      "Group_93.pdf\n",
      "Group_937.pdf\n",
      "Group_94.pdf\n",
      "Group_95.pdf\n",
      "Group_96.pdf\n",
      "Group_97.pdf\n",
      "Group_99.pdf\n"
     ]
    }
   ],
   "source": [
    "pdfs = listdir(out_dir)\n",
    "pdfs.sort()\n",
    "merger = PdfMerger()\n",
    "for pdf in pdfs:\n",
    "    if pdf.startswith('Group'):\n",
    "        merger.append(out_dir+'/'+pdf)\n",
    "        print(pdf)\n",
    "        \n",
    "merger.write(out_dir+'/Circos.pdf')\n",
    "merger.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
