{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from os import listdir, mkdir\n",
    "from os.path import isdir\n",
    "\n",
    "def Load_PAF(filepath):\n",
    "    lines = open(filepath).readlines()\n",
    "    header = ['Query','Qlen','QStart','QEnd','Orientation','Subject','SLen',\n",
    "              'SStart','SEnd','Matches','AlignLength','MAPQ']\n",
    "    op = []\n",
    "    for l in lines:\n",
    "        l = l.split('\\t')[:12]\n",
    "        op.append(dict(zip(header, l)))\n",
    "    df = pd.DataFrame(op)\n",
    "    df[['Qlen','QStart','QEnd','SLen','SStart',\n",
    "        'SEnd','Matches','AlignLength','MAPQ']] = df[['Qlen','QStart','QEnd','SLen','SStart',\n",
    "                                                      'SEnd','Matches','AlignLength','MAPQ']].astype('int')\n",
    "    df['PIdent'] = df['AlignLength']/df['Qlen']*100\n",
    "    df = df.loc[df.groupby(['Query'])['PIdent'].idxmax()]\n",
    "    return df\n",
    " \n",
    "def Assign_Coords(subg, filter_str):\n",
    "    g_undir = subg.to_undirected()\n",
    "    values, orderslist, osb_contigs_aligned = [], [], []\n",
    "    for n in subg.nodes():\n",
    "        try:\n",
    "            if filter_str in subg.nodes[n]['Memberships']:\n",
    "                osb_contigs_aligned.append(n)\n",
    "        except nx.NetworkXError:\n",
    "            continue\n",
    "    \n",
    "    for n in subg.nodes:\n",
    "        try:\n",
    "            if filter_str not in subg.nodes[n]['Memberships']:\n",
    "                neighbors = list(subg.predecessors(n)) + list(subg.successors(n))\n",
    "                cnt = len(set(neighbors).intersection(osb_contigs_aligned))\n",
    "                values.append(cnt)\n",
    "                orderslist.append(n)\n",
    "        except KeyError:\n",
    "            pass\n",
    "    orderslist = np.array(orderslist)[np.argsort(values)]\n",
    "    op = []\n",
    "    \n",
    "    for src in orderslist:\n",
    "        neighbors = g_undir.neighbors(src)\n",
    "        for n in neighbors:\n",
    "            flag = -1\n",
    "            if subg.nodes[n]['Orientation'] == '*':\n",
    "                continue\n",
    "            if subg.has_edge(n, src):\n",
    "                edge = (n, src)\n",
    "                edge_orientation = subg.edges[edge]['orientation']\n",
    "                edge_overlap = int(float(g_undir.edges[edge]['mean']))\n",
    "                c2_length = int(g_undir.nodes[src]['length'])\n",
    "                s1, e1 = subg.nodes[n]['Start'], subg.nodes[n]['End']\n",
    "                orientation = subg.nodes[n]['Orientation']\n",
    "                e1 = s1 + int(subg.nodes[n]['length'])\n",
    "                if orientation == '+' and subg.nodes[n]['orientation'] == 'REV':\n",
    "                    s1, e1 = e1, s1\n",
    "                if orientation == '-' and subg.nodes[n]['orientation'] == 'REV':\n",
    "                    s1, e1 = e1, s1\n",
    "                if edge_orientation == 'EE':\n",
    "                    e2 = e1 + edge_overlap\n",
    "                    s2 = e2 + c2_length\n",
    "                if edge_orientation == 'EB':\n",
    "                    s2 = e1 + edge_overlap\n",
    "                    e2 = s2 + c2_length\n",
    "                if edge_orientation == 'BB':\n",
    "                    s2 = s1 + edge_overlap\n",
    "                    e2 = s2 + c2_length\n",
    "                if edge_orientation == 'BE':\n",
    "                    e2 = s1 + edge_overlap\n",
    "                    s2 = e2 + c2_length\n",
    "                start,end = (s2, e2)\n",
    "                flag = 0\n",
    "            else:\n",
    "                edge = (src, n)\n",
    "                edge_orientation = subg.edges[edge]['orientation']\n",
    "                edge_overlap = int(float(g_undir.edges[edge]['mean']))\n",
    "                c1_length = int(g_undir.nodes[src]['length'])\n",
    "                s2, e2 = subg.nodes[n]['Start'], subg.nodes[n]['End']\n",
    "                e2 = s2 + int(subg.nodes[n]['length'])\n",
    "                orientation = subg.nodes[n]['Orientation']\n",
    "                if orientation == '+' and subg.nodes[n]['orientation'] == 'REV':\n",
    "                    s2, e2 = e2, s2\n",
    "                if orientation == '-' and subg.nodes[n]['orientation'] == 'REV':\n",
    "                    s2, e2 = e2, s2\n",
    "                if edge_orientation == 'EE':\n",
    "                    e1 = e2 - edge_overlap\n",
    "                    s1 = e1 - c1_length\n",
    "                if edge_orientation == 'EB':\n",
    "                    e1 = s2 - edge_overlap\n",
    "                    s1 = e1 - c1_length\n",
    "                if edge_orientation == 'BB':\n",
    "                    s1 = s2 - edge_overlap\n",
    "                    e1 = s1 - c1_length\n",
    "                if edge_orientation == 'BE':\n",
    "                    s1 = e2 - edge_overlap\n",
    "                    e1 = s1 - c1_length\n",
    "                start,end = (s1,e1)\n",
    "                flag = 1\n",
    "                \n",
    "            d = {'Contig':src, 'Parent_Node':n,'Start':start, 'End':end,\n",
    "                 'Membership':subg.nodes[src]['Memberships'],\n",
    "                 'Contig_Length':int(subg.nodes[src]['length']),\n",
    "                 'Parent_Length':int(subg.nodes[n]['length']), \n",
    "                 'OSA_PIdent(Contig)':subg.nodes[src]['OSA_PIdent'],\n",
    "                 'OSB_PIdent(Contig)':subg.nodes[src]['OSB_PIdent'],\n",
    "                 'OSA_PIdent(Parent)':subg.nodes[n]['OSA_PIdent'],\n",
    "                 'OSB_PIdent(Parent)':subg.nodes[n]['OSB_PIdent'],\n",
    "                 'Contig_Orientation(MetaCarvel)':subg.nodes[src]['orientation'],\n",
    "                 'Contig_Orientation(Minimap2)':subg.nodes[src]['Orientation'],\n",
    "                 'Parent_Orientation(MetaCarvel)':subg.nodes[n]['orientation'],\n",
    "                 'Parent_Orientation(Minimap2)':subg.nodes[n]['Orientation']}\n",
    "            if flag == 0:\n",
    "                d['Parent_Type'] = 'Parent'\n",
    "            elif flag == 1:\n",
    "                d['Parent_Type'] = 'Descendant'\n",
    "            op.append(d)\n",
    "    return op\n",
    "\n",
    "def Return_Memberships(df_osa, df_osb, G, genome, quality_cutoff = 80):\n",
    "    osa_contigs_aligned = df_osa[df_osa['PIdent'] >= quality_cutoff]['Query'].tolist()\n",
    "    osa_contigs_not_aligned = df_osa[df_osa['PIdent'] < quality_cutoff]['Query'].tolist()\n",
    "\n",
    "    osb_contigs_aligned = df_osb[df_osb['PIdent'] >= quality_cutoff]['Query'].tolist()\n",
    "    osb_contigs_not_aligned = df_osb[df_osb['PIdent'] < quality_cutoff]['Query'].tolist()\n",
    "    \n",
    "    df_osa_grp = df_osa[df_osa['PIdent'] >= quality_cutoff]\n",
    "    df_osa_grp = df_osa.set_index('Query').T.to_dict()\n",
    "    \n",
    "    df_osb_grp = df_osb[df_osb['PIdent'] >= quality_cutoff]\n",
    "    df_osb_grp = df_osb.set_index('Query').T.to_dict()\n",
    "    \n",
    "    d_memberships = {}\n",
    "    contigs = set(df_osa['Query'].tolist()).union(set(df_osb['Query'].tolist())).union(set(list(G.nodes())))\n",
    "    for c in contigs:\n",
    "        if c in osa_contigs_aligned:\n",
    "            try: d_memberships[c]['Memberships'].append('OSA')\n",
    "            except KeyError: d_memberships[c] = {'Memberships':['OSA']}\n",
    "\n",
    "        if c in osb_contigs_aligned:\n",
    "            try: d_memberships[c]['Memberships'].append('OSB')\n",
    "            except KeyError: d_memberships[c] = {'Memberships':['OSB']}\n",
    "\n",
    "        if c not in osa_contigs_aligned and c not in osb_contigs_aligned:\n",
    "            d_memberships[c] = {'Memberships' : []} \n",
    "        \n",
    "        try: d_memberships[c]['OSA_PIdent'] = df_osa_grp[c]['PIdent']\n",
    "        except KeyError: d_memberships[c]['OSA_PIdent'] = 0\n",
    "        \n",
    "        try: d_memberships[c]['OSB_PIdent'] = df_osb_grp[c]['PIdent']\n",
    "        except KeyError: d_memberships[c]['OSB_PIdent'] = 0\n",
    "        \n",
    "        if genome == 'OSA':\n",
    "            try:\n",
    "                Start = min(df_osa_grp[c]['SStart'], df_osa_grp[c]['SEnd'])\n",
    "                End = max(df_osa_grp[c]['SStart'], df_osa_grp[c]['SEnd'])\n",
    "                Orientation = df_osa_grp[c]['Orientation']\n",
    "                d_memberships[c]['Start'] = Start\n",
    "                d_memberships[c]['End'] = End\n",
    "                d_memberships[c]['Orientation'] = Orientation\n",
    "            except KeyError:\n",
    "                d_memberships[c]['Start'] = -1\n",
    "                d_memberships[c]['End'] = -1\n",
    "                d_memberships[c]['Orientation'] = '*'\n",
    "                \n",
    "        if genome == 'OSB':\n",
    "            try:\n",
    "                Start = min(df_osb_grp[c]['SStart'], df_osb_grp[c]['SEnd'])\n",
    "                End = max(df_osb_grp[c]['SStart'], df_osb_grp[c]['SEnd'])\n",
    "                Orientation = df_osb_grp[c]['Orientation']\n",
    "                d_memberships[c]['Start'] = Start\n",
    "                d_memberships[c]['End'] = End\n",
    "                d_memberships[c]['Orientation'] = Orientation\n",
    "            except KeyError:\n",
    "                d_memberships[c]['Start'] = -1\n",
    "                d_memberships[c]['End'] = -1\n",
    "                d_memberships[c]['Orientation'] = '*'\n",
    "    nx.set_node_attributes(G, d_memberships)\n",
    "    return G\n",
    "\n",
    "def Main(osa_path, osb_path, Graph_path, genome, quality_cutoff):\n",
    "    df_osa = Load_PAF(osa_path)\n",
    "    df_osb = Load_PAF(osb_path)\n",
    "\n",
    "    G = nx.read_gml(Graph_path)\n",
    "    G = Return_Memberships(df_osa, df_osb, G, genome, quality_cutoff = quality_cutoff)\n",
    "    \n",
    "    ctr = 0\n",
    "    oplist = []\n",
    "    \n",
    "    for c in G.nodes():\n",
    "        neighbors = list(G.predecessors(c)) + list(G.successors(c))\n",
    "        found = False\n",
    "\n",
    "        for n in neighbors:\n",
    "            try:\n",
    "                if genome not in G.nodes[n]['Memberships']:\n",
    "                    found = True\n",
    "                    break\n",
    "            except KeyError:\n",
    "                continue\n",
    "\n",
    "        if found == True:\n",
    "            subg = G.subgraph([c]+neighbors)\n",
    "            op = Assign_Coords(subg, genome)\n",
    "            oplist += op\n",
    "            ctr+=1\n",
    "    \n",
    "    df_scaffold = pd.DataFrame(oplist)\n",
    "    df_scaffold['Membership'] = df_scaffold['Membership'].astype(str)\n",
    "    df_scaffold = df_scaffold.drop_duplicates()\n",
    "    return df_scaffold\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Sort the contigs based on their start coordinate. \n",
    "2. Starting at the first contig, assign coordinates to its putative descendants based on all its neighbors found in the reference. \n",
    "3. The coordinates must respect the orientation, length and overlaps. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pardir = '/Users/harihara/Mount-2/hotspring_metagenome/Synechococcus_paper_analysis/reassembly/'\n",
    "files = listdir(pardir)\n",
    "if not isdir('/Users/harihara/Research-Activities/Data/Hot-Spring/Missing_Contig_Coords_2/'):\n",
    "    mkdir('/Users/harihara/Research-Activities/Data/Hot-Spring/Missing_Contig_Coords_2/')\n",
    "if not isdir('/Users/harihara/Research-Activities/Data/Hot-Spring/Missing_Contig_Coords_2/OSA/'):\n",
    "    mkdir('/Users/harihara/Research-Activities/Data/Hot-Spring/Missing_Contig_Coords_2/OSA/')\n",
    "if not isdir('/Users/harihara/Research-Activities/Data/Hot-Spring/Missing_Contig_Coords_2/OSB/'):\n",
    "    mkdir('/Users/harihara/Research-Activities/Data/Hot-Spring/Missing_Contig_Coords_2/OSB/')\n",
    "\n",
    "files = ['HotsprottomLayer_megahit_assembled_contigs_osa']\n",
    "\n",
    "for f in files:\n",
    "    if isdir(pardir+f) and ('osa' in f or 'osb' in f):\n",
    "        if 'osa' in f:\n",
    "            sample_id = f.replace(\"_megahit_assembled_contigs_osa\",\"\")\n",
    "            osa_path = pardir+'contig_mapping_paf/'+sample_id+'_osa_contigs_aligned_to_osa.paf'\n",
    "            osb_path = pardir+'contig_mapping_paf/'+sample_id+'_osa_contigs_aligned_to_osb.paf'\n",
    "            g_path =  pardir+f+'/'+sample_id+'_osa_scaffolds/oriented.gml'\n",
    "            df_scaffold = Main(osa_path, osb_path, g_path, 'OSA', 80)\n",
    "            print(f,len(df_scaffold))\n",
    "            df_scaffold.to_csv('/Users/harihara/Research-Activities/Data/Hot-Spring/Missing_Contig_Coords_2/OSA/'+sample_id+'.txt', \n",
    "                               sep = \"\\t\")\n",
    "            print(df_scaffold[df_scaffold['Contig'] == 'k141_1518'])\n",
    "            \n",
    "        if 'osb' in f:\n",
    "            sample_id = f.replace(\"_megahit_assembled_contigs_osb\",\"\")\n",
    "            osa_path = pardir+'contig_mapping_paf/'+sample_id+'_osb_contigs_aligned_to_osa.paf'\n",
    "            osb_path = pardir+'contig_mapping_paf/'+sample_id+'_osb_contigs_aligned_to_osb.paf'\n",
    "            g_path = pardir+f+'/'+sample_id+'_osb_scaffolds/oriented.gml'\n",
    "            df_scaffold = Main(osa_path, osb_path, g_path, 'OSB', 80)\n",
    "            print(f,len(df_scaffold))\n",
    "            df_scaffold.to_csv('/Users/harihara/Research-Activities/Data/Hot-Spring/Missing_Contig_Coords_2/OSB/'+sample_id+'.txt', \n",
    "                               sep = \"\\t\")\n",
    "            print(df_scaffold[df_scaffold['Contig'] == 'k141_1518'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
